### ISSUE TITLE
Legacy and Split UPE flows: No metadata for setup intents

### ISSUE THREAD

**Issue by htdat** [Automattic/woocommerce-payments#6575]
### Describe the bug

Currently, no metadata is attached to setup intents in Legacy and Split UPE flows.

### To Reproduce

Purchase a subscription with a free trial to see this issue.

0. Enable legacy or split UPE flow.
1. Use WCPay Subscription and create a subscription product with free trial.
2. Purchase this product, and check out with a new card.
3. Check the setup intent attached to this order.
4. Error: There is no metadata attached to this setup intent.

### Actual behavior

Metadata for setup intents should be added, which is similar to payment intents as well as the traditional card flow.

### Dev note 

I think that's because during `payment_process` of UPE, only [PaymentIntent (i.e. `$payment_needed === true`)](https://github.com/Automattic/woocommerce-payments/blob/1ff019d228efcb6c5c3e1730a47caf764e025d2c/includes/payment-methods/class-upe-payment-gateway.php#L474-L575) is updated with metadata.

For [the traditional (card) flow](https://github.com/Automattic/woocommerce-payments/blob/1ff019d228efcb6c5c3e1730a47caf764e025d2c/includes/class-wc-payment-gateway-wcpay.php#L1066-L1233), metadata is added for both [PaymentIntent](https://github.com/Automattic/woocommerce-payments/blob/1ff019d228efcb6c5c3e1730a47caf764e025d2c/includes/class-wc-payment-gateway-wcpay.php#L1111-L1117) and [SetupIntent](https://github.com/Automattic/woocommerce-payments/blob/1ff019d228efcb6c5c3e1730a47caf764e025d2c/includes/class-wc-payment-gateway-wcpay.php#L1216-L1219). 

We would check what the code has done for PaymentIntent and consider doing the same for SetupIntent. 

### Additional context

- See SetupIntents on Stripe between two flows: 30d23-pb/#plain
- On-going discussion: p1687513417167319-slack-C05AW31EKFF

**Comment by timur27** [comment]
In the legacy gateway, we [create and confirm setup intent at once](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L1216). [We do the same for the payment intents](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L1111). Metadata is part of the request class, and things work smoothly for the legacy gateway.

For the UPE, we have creation and confirmation/update steps split in two for obvious reasons - legacy UPE/split UPE require payment intents to be created to render the Payment Element. Thus, **_for payment intents_**, we [create an intent](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L329) separately, without providing any meta-object, and then after users click _Place order_, we [update the intent](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L518) by providing more details to it.

The problem is that for the **_setup intents_**, only [intent creation](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L413) is implemented. There's simply no implementation that would [update setup intent](https://stripe.com/docs/api/setup_intents/update). 

To fill this gap, we'd need to:
1. Create a new object similar to [`Update_Intention` request class](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/core/server/request/class-update-intention.php). Based on some discussions and my understanding, we can't re-use or modify this request class for our current purposes, since we took the path of treating each action separately with a separate command-class responsible for it and it would make sense to continue following this path. Moreover, `Update_Intention` has `amount` validation which excludes the possibility of creating a request with zero amount.
2. Introduce the implementation that handles the newly created request class, similarly to the implementation e.g., for `Update_Intention` request handling.
3. Add an `else` block following [this if which handles payment intents](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L475C11-L475C11). `Else` block will eventually be our new home for update-setup-intent-related logic.
4. Include logic similar to [this](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L518-L538) into the `else` block, which will build a request, send it to the server, and the response will be later used for https://github.com/Automattic/woocommerce-payments/issues/6576's purposes. 
5. Maybe think about renaming [`Update_Intention`](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/core/server/request/class-update-intention.php) class to emphasize that it's for _payment_ intents only. Right now, it's acceptable for this class to have such a name, but once the setup intent request class is introduced, we might want to specify a bit more.

**Comment by htdat** [comment]
Excellent recap! I totally agree with this redirection. cc @Automattic/gamma - what do you think, especially bullet 1 as it's related to core request classes?

> Maybe think about renaming [Update_Intention](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/core/server/request/class-update-intention.php) class to emphasize that it's for payment intents only. Right now, it's acceptable for this class to have such a name, but once the setup intent request class is introduced, we might want to specify a bit more.

Yes, we can do this as a follow-up issue.

**Comment by RadoslavGeorgiev** [comment]
> what do you think, especially bullet 1 as it's related to core request classes?

I was the one to suggest creating another request class, and the endpoint on the server, so I agree with myself :D

Anyway, do you think there is any alternative @htdat ?

**Comment by htdat** [comment]
@RadoslavGeorgiev :D No alternative from my side. I am also just curious if there is any other approach.

**Comment by htdat** [comment]
Hi @timur27, as Rado and I discussed above, your proposal sounds good. Would your team (cc @FangedParakeet too as I see you also involved in these issues) take it and implement the fix?

**Comment by FangedParakeet** [comment]
Greetings one and all! First of all, just want to affirm my accordance with everything @timur27 stated [above](https://github.com/Automattic/woocommerce-payments/issues/6575#issuecomment-1604444605). I'd also agree that, while not necessarily required for rollout, closing this issue would probably be ancillary to our goals to use the deferred intent UPE to complete our gateway as the chosen one to consolidate, assimilate, and succeed them all. Although, I'd likely not currently consider this work antecedent to anything else that we are currently in the midst of. 

Have we decided any tentative priority towards resolving this issue? It doesn't seem necessarily tied to nor blocking any ongoing security work and since it will require changes on both the plugin and server, it seems like we may want to take a more laboured and measured approach towards ensuring the solidity of our eventual solution.

In short, I'd welcome adding this to our [project work](https://github.com/Automattic/woocommerce-payments/issues/5019), but would probably leave it in the backlog for now, unless there is any urgency to complete this within the current sprint instead.

**Comment by htdat** [comment]
@FangedParakeet - that would work for us. 

Just FYI. Until this issue is resolved, it will be still a blocker for Gamma's another PR https://github.com/Automattic/woocommerce-payments/pull/6487 that is in the "ready to merge" state, which uses order_id attached in the order meta of both payment and setup intents

**Comment by FangedParakeet** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Heisenberg-63586f2ec60359001310fa6a/issues/Automattic/woocommerce-payments/6575?planning-poker) with Zenhub @timur27

**Comment by FangedParakeet** [comment]
@htdat, just discussed this with @timur27 today and we've had a slight change of heart on the issue here. This issue only affects the legacy UPE and current split UPE; crucially it will **not** affect the deferred intent UPE implementation whose rollout looms on the horizon. 

If we were to prioritise this work now, as stated above, we would probably only commence resolving this issue _after_ releasing the deferred intent UPE. Moreover the rollout of the new deferred intent UPE will almost certainly be followed by a swift deprecation of the legacy UPE (see paJDYF-8UK-p2#comment-19004), such that it doesn't enormously appear to be  in our interest to continue supporting functionality for the deprecated legacy/split UPE. Consequently, I am now hesitant to commit to this work now, since the deferred intent UPE will both resolve this issue and make any solution implemented here rapidly redundant. 

> it will be still a blocker for Gamma's another PR https://github.com/Automattic/woocommerce-payments/pull/6487

It looks like #6487 will use the stored metadata to prevent duplicate orders. Would leaving this issue open specifically block that PR or would it just mean that that PR would be unable to prevent duplicate orders where free-trial subscriptions are concerned? If the latter case is true, that seems more like an unresolved edge-case rather than an absolute blocker, but correct me if I am mistaken.

Please let me know how any of that sounds. If you feel that this issue is indeed critical enough that it should be solved before completing our implementation of the deferred intent UPE, then we can give this another look and try to prioritise it a little earlier. Although, as things currently stand, since this issue will require robust changes on both the plugin and server, I foresee that closure over this issue would likely intersect with our rollout of this new feature, in which case by the time the solution is completed it will likely already be redundant. Cheers!

**Comment by htdat** [comment]
Thanks for the context @FangedParakeet. 

> It looks like https://github.com/Automattic/woocommerce-payments/pull/6487 will use the stored metadata to prevent duplicate orders. Would leaving this issue open specifically block that PR or would it just mean that that PR would be unable to prevent duplicate orders where free-trial subscriptions are concerned? If the latter case is true, that seems more like an unresolved edge-case rather than an absolute blocker, but correct me if I am mistaken.

Without the fix for this issue, it would be that we can not prevent duplicate orders for free-trial subscription orders. But I will check if with Gamma and see if we can either (1) fix this issue, (2) iterate #6487 to fix for only payment intents. 

_Update_: we chose this path: "(2) iterate #6487 to fix for only payment intents", and merged the PR.

**Comment by frosso** [comment]
The implementation of the Deferred Intent Split UPE will fix this issue automatically. Implementing a fix on the Split UPE or Legacy UPE (not Deferred Intent) would require more work and wouldn't be a good ROI.

We will close this issue in favor of migrating to the Deferred Intent Split UPE.
